### ISSUE TITLE
Break out WebSocket into its own module/repo?

### ISSUE THREAD

**Issue by blesh** [ReactiveX/rxjs#1278]
Should we break WebSocketSubject into it's own package and repository? There aren't many good reasons to keep it here, and this would enable separate versioning of the two. Then again there aren't many good reasons _not_ to keep it here.

**Comment by trxcllnt** [comment]
@blesh https://github.com/babel/babel/blob/master/doc/design/monorepo.md

**Comment by blesh** [comment]
@trxcllnt this is mostly being done at the request of @mattpodwysocki ... I'm relatively indifferent. 

FWIW: Projects like Ember and Angular 2 have one repository but publish many packages. We're already doing some of this (with rxjs and rxjs-es). So the next question is: Is it worth publishing a separate package for a web socket or ajax?

- Do Node folks care about DOM functionality? Probably not, but maybe.
- Is node_module bloat even a thing reasonable people care about? I dunno. I know I don't, but there could be valid concerns I'm not aware of.

**Comment by david-driscoll** [comment]
Normally I'm not in favor of the mono-repo but in this case there could be some advantages.  I'm not a core team member, and really I'll still contribute no matter what direction the Rx team takes. :smile: 


Pros
* With everything together it would be easy to see all the source code (especially with shared code)
* Versions would be in sync with the parent rxjs module, which would avoid potential breaking changes.
* Keeps tests together, to watch for breaking changes.


Cons
* Versions rev faster on components that don't actually change (but you'd be getting new versions of Rx which is good!).
* Something like `npm link` may be a pain for developers trying to do local development with the linked package.
  * This would be mitigated if the folders were mapped similar to babels
    * `/packages/rxjs`
    * `/packages/rxjs-dom`
    * `/packages/rxjs-ws`

**Comment by blesh** [comment]
Actually, I take back what I said about Ember. @jayphelps pointed out that Ember doesn't do this, so that doc that @trxcllnt linked above isn't accurate WRT that.

**Comment by masaeedu** [comment]
:+1: for this, other people writing and distributing their own operators should be a supported use-case, so this will be a good training run for that. Would distributing the websocket code as a separate npm package dependent on Rx not work?

**Comment by rgbkrk** [comment]
I'm not worried about the size of `node_modules` here, especially since I'm using the `@reactivex/rxjs` install. Took me a while to figure out the export paths with TS while writing up #1410.

**Comment by jayphelps** [comment]
FWIW I vote monorepo. There are numerous reasons to keep it in one repo, mostly having a single build/test/publish pipeline, while keeping different npm packages. That means that npm bloat doesn't happen because the npm artifacts only include the items for their package.

Whether it's worth separating `AjaxObservable` and `WebSocketSubject` instead of just a `rxjs-dom` package is another topic. It's certainly possible but I don't think the effort is worth it. I don't feel strongly either way, though.

```
/src
    /rxjs
        package.json (@reactivex/rxjs)
    /rxjs-dom
        package.json (@reactivex/rxjs-dom)
```

**Comment by staltz** [comment]
:+1: monorepo dealing with different npm packages

**Comment by meriturva** [comment]
+1

**Comment by blesh** [comment]
Closed in favor of #1506
